[ES|QL] Date nanos implicit casting in union types

fang-xing-esql opened the pull request:
Resolves #110009 

The goal of this PR is to support union typed fields that reference both `date` and `date_nanos` fields without explicit casting. Union typed(mixed with `date` and `date_nanos`) fields are casted to a common data type during Analyzer phase automatically. `numeric` or `string` typed fields are not within the scope of this PR.

Today, queries that reference union(mixed with `date` and `date_nanos`) typed fields like below either return nulls or fail, because they reference union typed fields without explicit casting. After this PR, they should not fail or return null.

```
mapping
curl -u elastic:password -X PUT "localhost:9200/sample_data_1?pretty" -H 'Content-Type: application/json' -d'
{
  "mappings": {
    "properties": {
      "client.ip": {"type": "ip"},
      "message": {"type": "text"},
      "test_date": {"type": "date"},
      "event.duration": {"type": "long"}
    }
  }
}
'
curl -u elastic:password -X PUT "localhost:9200/sample_data_2?pretty" -H 'Content-Type: application/json' -d'
{
  "mappings": {
    "properties": {
      "client.ip": {"type": "ip"},
      "message": {"type": "keyword"},
      "test_date": {"type": "date_nanos"},
      "event.duration": {"type": "integer"}
    }
  }
}
'
queries
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "FROM sample_data*"
}
'
       @timestamp       |   client.ip   |event.duration |    message    |   test_date   
------------------------+---------------+---------------+---------------+---------------
2023-10-23T12:15:03.360Z|172.21.2.162   |null           |null           |null           
2023-10-23T12:27:28.948Z|172.21.2.113   |null           |null           |null           
2023-10-23T13:33:34.937Z|172.21.0.5     |null           |null           |null           
2023-10-23T13:51:54.732Z|172.21.3.15    |null           |null           |null           
2023-10-23T13:52:55.015Z|172.21.3.15    |null           |null           |null           
2023-10-23T13:53:55.832Z|172.21.3.15    |null           |null           |null           
2023-10-23T13:55:01.543Z|172.21.3.15    |null           |null           |null           
2023-10-23T12:15:03.360Z|172.22.2.162   |null           |null           |null           
2023-10-23T12:27:28.948Z|172.22.2.113   |null           |null           |null           
2023-10-23T13:33:34.937Z|172.22.0.5     |null           |null           |null           
2023-10-23T13:51:54.732Z|172.22.3.15    |null           |null           |null           
2023-10-23T13:52:55.015Z|172.22.3.15    |null           |null           |null           
2023-10-23T13:53:55.832Z|172.22.3.15    |null           |null           |null           
2023-10-23T13:55:01.543Z|172.22.3.15    |null           |null           |null           

+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "FROM sample_data* | sort test_date"
}
'
{
  "error" : {
    "root_cause" : [
      {
        "type" : "verification_exception",
        "reason" : "Found 1 problem\nline 1:26: Cannot use field [test_date] due to ambiguities being mapped as [2] incompatible types: [date_nanos] in [sample_data_2], [datetime] in [sample_data_1]"
      }
    ],
    "type" : "verification_exception",
    "reason" : "Found 1 problem\nline 1:26: Cannot use field [test_date] due to ambiguities being mapped as [2] incompatible types: [date_nanos] in [sample_data_2], [datetime] in [sample_data_1]"
  },
  "status" : 400
}
```

A field can be referenced in nearly all the processing commands, and this PR should cover all the possible places a (union typed)field can appear:

- [x] evalCommand
- [x] whereCommand
- [x] keepCommand
- [x] sortCommand
- [x] dropCommand
- [x] renameCommand
- [x] statsCommand
- [x] dissectCommand - works for keyword and text only, out of the scope of this PR
- [x] grokCommand - works for keyword and text only, out of the scope of this PR
- [x] enrichCommand - enrich does not support date_nanos yet
- [x] mvExpandCommand
- [x] joinCommand - lookup join on date_nanos field does not work today, out of the scope of this PR

fang-xing-esql committed:
implicit casting for date nanos

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
implicit casting for union typed field in logical plan

elasticsearchmachine commented:
Hi @fang-xing-esql, I've created a changelog YAML for you.

fang-xing-esql committed:
Update docs/changelog/123678.yaml

fang-xing-esql commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
more implicit casting for union typed field in logical plan

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
refactor

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
support aggregations with bucket and more tests

fang-xing-esql committed:
correct typo

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

costin commented:
Fang, can you please double check the status of this PR? Is it ready for review or still working on it?

fang-xing-esql commented:
> Fang, can you please double check the status of this PR? Is it ready for review or still working on it?

I'm trying to make enrich work with union typed fields, and this is the last piece that I want to figure out before making this ready for review.

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
remove grok and dissect, add enrich

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
more tests around date functions

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -1639,3 +1639,472 @@ id:integer | name:keyword | count:long
+| limit 2
+;
+
+x:long |languages:integer |hire_date:date_nanos     |avg_worked_seconds:unsigned_long
```
Why do we expand the scope of this PR to include autocasting of long/int multi-typed fields to long?

I thought the date_nanos/datetime autocasting was special because it avoids problems with `@timestamp` fields that may change types as the index evolves, which justifies autocasting in this one case - whereas we generally decided against auto casting in favor of explicit casting.

What is the target scale of autocasted types that we aim to achieve? What are potential side effects from increasing the scope from just autocasting datetime/date_nanos to other type combinations?

@craigtaverner , do you recall why we specifically had union types not auto cast in the initial implementation?

fang-xing-esql replied:
Implicit casting fro `numeric` union typed fields is removing from this PR.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -1639,3 +1639,472 @@ id:integer | name:keyword | count:long
+
+FROM employees, employees_incompatible
+| DROP birth_date
+| EVAL x = emp_no + 1, y = hire_date + 1 year, z = hire_date::datetime
```
What happens to compute `hire_date::datetime`? Is all converted to nanos, first, when extracting the field initially - and then cast to datetimes?

If so, this would be losing datetimes that are far in the future/past. We should probably ensure that we skip the first cast to date_nanos if the first thing we do with the field attribute is casting it back to datetime. If we don't do that already. (Not deep enough into the review, yet, so apologies if that's already addressed.)

fang-xing-esql replied:
We don't double cast, if a (union typed)field has been cast to a type explicitly, it won't be casted to date_nanos implicitly again. `ResolveUnionTypes` transforms the `InvalidMappedField`s(resolved) with explicit casting into `MultiTypeEsField`s, and `ImplicitCastingForUnionTypedFields` happens after it. `ImplicitCastingForUnionTypedFields` finds the `InvalidMappedField`s(resolved, mixed with `date` and `date_nanos`) without explicit casting, and adds implicit casting for them, the `AnalyzerTests.testImplicitCastingForDateAndDateNanosFields` is to check there is no double casting.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
I also wonder:

For wide index pattern queries like the simple `FROM logs-*`, currently the result on large clusters with many indices consists of maaaany null columns, because of mapping conflicts.

Loading/sending constant null blocks is comparably cheap.

If we start enabling auto-casting whenever we can, these queries will now read and return actual data; this might cause performance issues for this kind of query, no?

@fang-xing-esql , maybe we should check in with the others, ensure that this case is covered by our nightlies, and observe if this noticeably changes the performance in some cases, once this is merged?

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -1639,3 +1639,472 @@ id:integer | name:keyword | count:long
+required_capability: implicit_casting_union_typed_numeric_and_date
+
+FROM employees, employees_incompatible
+| EVAL language_code = languages
```
Nit: not sure the `languages` is an interesting field; it's integer in one index and byte in the other, which are internally represented as integer anyway. It'd be more interesting to perform a lookup on a long that would've failed before this PR.

fang-xing-esql replied:
`enrich` and `lookup join` do not support data_nanos as join fields yet, and `numeric` is out of the scope, this test is removed.

alex-spies reviewed (commented):
I'm not done with the review yet, but left a first round of remarks - mostly questions.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into date-nanos-implicit-casting

alex-spies reviewed (changes requested):
Ok, done now. Thanks @fang-xing-esql , this is a good start and I like the strong test coverage.

However, I see the following problems:
- This seems to go straight to production. Please, let's guard this behind SNAPSHOT for now. We could add a setting or pragma to control this, and flip the setting's default value later when we're confident that all is well.
- The scope is expanded to all numerical types. I think we should check back that this really doesn't have unintended side effects before moving forward.
- This could accidentally worsen the performance of wide index pattern queries, like `FROM log-s* | limit 10`. We'd have to confirm via benchmarks.
- I think the scope is also accidentally expanded to allow arbitrary comparisons between date time and nanos in any expression, due to the change to the EsqlDataTypeConverter, but we don't have corresponding tests.
- The code needs to special case, in multiple places, for a white list of commands that are currently working and tested with this. This is brittle and will break when we add new commands. This also adds a lot of code to facilitate this separately from the ResolveUnionTypes analyzer rule by injecting Evals, which I think could be simplified. We also require changes to the type resolution of several commands to account for UNSUPPORTED, which we didn't need for union types, previously.

I think the injecting of EVALs also implies that in a query like
```
FROM idx* | eval x  =  multi_typed_date_and_nanos | eval y = multi_typed_date_and_nanos::datetime
```
the column `y` will be computed by first transforming to nanos on block loading, and then to datetimes in the compute engine - rather than loading the blocks as datetimes directly, leading to a lossy conversion.

I'm not sure, but the same could happen between long and float/double, potentially leading to unintended lossy conversions if the long has more digits than float/double can hold?

I think there may be a simpler way to implement this, which would lean on the existing union type logic more and wouldn't have to special case so much, or require injecting Evals, or run into the problem of accidental lossy conversions.
Currently, in ResolveUnionTypes, we [scan the query for expressions](https://github.com/elastic/elasticsearch/blob/1f68bfbc3e1f3c0c87b0ad176577d17046be79fd/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java#L1581-L1583) of the kind `some_multi_typed_field::target_type` -  and if we find them, we create a new field attribute in the EsRelation for the field loaded as the target type, and replace the expression by this field attribute. (We also deduplicate based on the target type if we've already come across this conversion).
I expect that we could get a simpler implementation if we expanded the scope of the rule to also look for multi-typed field attributes (except those that were already replaced because they were inside a convert function) and _treat those field attributes as if they were of the form `some_multi_typed_field::target_type`_, and just apply the same substitution by the field attribute that's been cast to the target type.
If I'm correct, the changes may be reduced to just determining which target type to apply there - everything else should work as it always did with union types.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1346,6 +1379,232 @@ private static Expression cast(org.elasticsearch.xpack.esql.core.expression.func
+                        for (Expression e : newProjections) { // Add groupings
+                            newAggs.add(Expressions.attribute(e));
+                        }
+                        return agg.with(evalForInvalidMappedField(agg.source(), agg.child(), aliases), newProjections, newAggs);
```
Because we inject an `EVAL` here, there's a chance that we'll accidentally inject `EVAL`s multiple times. Which could lead to broken plans, when an upstream `EVAL` already replaced the multi-typed field attribute by a reference attribute of the same name. Or, put the other way around, since we're replacing a field by another one with the same name, downstream plans may have invalid references now.

This could happen in case that the reference resolution manages to resolve the whole query plan, with multiple references to multi-typed fields, before we start adding Evals.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1346,6 +1379,232 @@ private static Expression cast(org.elasticsearch.xpack.esql.core.expression.func
+                        List<? extends NamedExpression> origAggs = agg.aggregates();
+                        List<NamedExpression> newAggs = new ArrayList<>(origAggs.size());
+                        for (int i = 0; i < origAggs.size() - newProjections.size(); i++) { // Add aggregate functions
+                            newAggs.add(origAggs.get(i));
```
Normally, the way to update the expressions of a plan is to transform the expressions and use `AttributeMap.resolve` with an `AttributeMap<Attribute>` that contains which attribute should be replaced by which other attribute.

craigtaverner committed:
DateNanos should now require a minimum of three decimal places

A date in the valid date-nanos range should be acceptable to parse. This issue was showing up in a difference in behaviour between Elasticsearch ingest and the CsvTests. This fix makes CsvTests behave the same as Elasticearch itself.

craigtaverner reviewed (commented):
Ran out of time, so I'll post the comments so far.

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java:
```diff
@@ -947,7 +947,12 @@ public enum Cap {
+        /**
+         * Support implicit casting for union typed numeric and date/date_nanos fields
+         */
+        IMPLICIT_CASTING_UNION_TYPED_NUMERIC_AND_DATE;
```
I'll put it under snapshot.

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Eval.java:
```diff
@@ -132,7 +131,12 @@ private List<Alias> renameAliases(List<Alias> originalAttributes, List<String> n
 
     @Override
     public boolean expressionsResolved() {
-        return Resolvables.resolved(fields);
```
I was surprised that we don't always check if the data type before marking a `LogicalPlan` as resolved, as we do for an `Expression`, until I found we did it for [Rename](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Rename.java#L61), so I added them here for these commands, perhaps it is unintended, and because we never hit problems with data types before.

craigtaverner replied:
Is this code still required after the new approach in Analyzer?

fang-xing-esql commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/data/employees_incompatible.csv:
```diff
@@ -1,101 +1,101 @@
-1961-09-23T00:00:00Z,10098,Sreekrishna,F,1985-05-13T00:00:00Z,4,4,4,4,Servieres   ,44817,2.00,2.00,2.00,2.00,false,272392146,[Architect,Internship,Senior Team Lead],[false],[-2.83,8.31,4.38],[-2,8,4],[-2,8,4],[-2.83,8.31,4.38]
-1956-05-25T00:00:00Z,10099,Valter     ,F,1988-10-18T00:00:00Z,2,2,2,2,Sullins     ,73578,1.81,1.81,1.81,1.81,true ,377713748,[],[true,true],[10.71,14.26,-8.78,-3.98],[10,14,-8,-3],[10,14,-8,-3],[10.71,14.26,-8.78,-3.98]
-1953-04-21T00:00:00Z,10100,Hironobu   ,F,1987-09-21T00:00:00Z,4,4,4,4,Haraldson   ,68431,1.77,1.77,1.77,1.77,true ,223910853,[Purchase Manager],[false,true,true,false],[13.97,-7.49],[13,-7],[13,-7],[13.97,-7.49]
+birth_date:date     ,emp_no:long,first_name:text,gender:text,hire_date:date_nanos,languages:byte,languages.long:long,languages.short:short,languages.byte:byte,last_name:text,salary:long,height:float,height.double:double,height.scaled_float:scaled_float,height.half_float:half_float,still_hired:keyword,avg_worked_seconds:unsigned_long,job_positions:text,is_rehired:keyword,salary_change:float,salary ...
```
In the [mapping](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/qa/testFixtures/src/main/resources/mapping-default-incompatible.json#L16) , `birth_date` is defined as `date`, but not `date_nanos`, the header of this file doesn't match the mapping, and the values of the `birth_date` field are not within valid date_nanos range, that's the reason why the header is changed. `hire_date` is defined as `date_nanos` in the mapping, and they are in the range of valid `date_nanos`.

fang-xing-esql commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/FieldExtractorTestCase.java:
```diff
@@ -1002,6 +1005,9 @@ public void testLongIntegerConflict() throws IOException {
      * In an ideal world we'd promote the {@code short} to an {@code integer} and just go.
      */
     public void testIntegerShortConflict() throws IOException {
+        if (EsqlCapabilities.Cap.IMPLICIT_CASTING_UNION_TYPED_NUMERIC_AND_DATE.isEnabled()) {
+            return;
+        }
```
I wasn't very sure if we should do implicit casting for mixed numeric fields, if we decided to do it, these tests should be changed, instead of being skipped.

fang-xing-esql commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/data/employees_incompatible.csv:
```diff
@@ -1,101 +1,101 @@
-birth_date:date_nanos     ,emp_no:long,first_name:text,gender:text,hire_date:date_nanos,languages:byte,languages.long:long,languages.short:short,languages.byte:byte,last_name:text,salary:long,height:float,height.double:double,height.scaled_float:scaled_float,height.half_float:half_float,still_hired:keyword,avg_worked_seconds:unsigned_long,job_positions:text,is_rehired:keyword,salary_change:float, ...
```
Answered [here](https://github.com/elastic/elasticsearch/pull/123678/files#r2045346380).

craigtaverner replied:
The values in this column are not valid date_nanos, so unless we also want to change all values, we need to keep the type date. We have the hire_date column as a valid column for testing date_nanos, because the values are within range.

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
refactor and remove implicit casting for numeric types

fang-xing-esql commented on docs/changelog/123678.yaml:
```diff
@@ -0,0 +1,6 @@
+pr: 123678
+summary: Date nanos implicit casting
```
Implicit casting `numeric` union typed fields are removed from this PR.

fang-xing-esql committed:
update tests

fang-xing-esql committed:
fix tests

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java:
```diff
@@ -404,6 +404,9 @@ public static DataType commonType(DataType left, DataType right) {
             }
+            if ((isDateTime(left) && right == DATE_NANOS) || (left == DATE_NANOS && isDateTime(right))) {
+                return DATE_NANOS;
+            }
```
This is moved to `Analyzer`, as we limit the scope of this PR to `date` and `date_nanos` only, this `commpnType` method will not be called.

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1145,7 +1155,8 @@ private LogicalPlan resolveEnrich(Enrich enrich, List<Attribute> childrenOutput)
                     DataType[] allowed = allowedEnrichTypes(matchType);
-                    if (Arrays.asList(allowed).contains(dataType) == false) {
+                    // leave multi-typed fields to ImplicitCasting to cast to a common type
+                    if (Arrays.asList(allowed).contains(dataType) == false && multiTypedField(resolved) == null) {
```
enrich does not support on date_nanos fields, this changed was added for `numeric` type fields, as `numeric` types  are removed from this PR, this change is not needed.

fang-xing-esql commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -133,8 +133,8 @@ public void testUnsupportedAndMultiTypedFields() {
             error("from test* | enrich client_cidr on unsupported", analyzer)
         );
         assertEquals(
-            "1:36: Unsupported type [unsupported] for enrich matching field [multi_typed];"
```
The change in `resolveEnrich` is removed, so there is no need to change this test now.

fang-xing-esql commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/FieldExtractorTestCase.java:
```diff
@@ -1002,6 +1005,9 @@ public void testLongIntegerConflict() throws IOException {
      * In an ideal world we'd promote the {@code short} to an {@code integer} and just go.
      */
     public void testIntegerShortConflict() throws IOException {
+        if (EsqlCapabilities.Cap.IMPLICIT_CASTING_UNION_TYPED_NUMERIC_AND_DATE.isEnabled()) {
```
After taking out `numeric` types from this PR, we don't need to change this test any more.

fang-xing-esql commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -1639,3 +1639,472 @@ id:integer | name:keyword | count:long
+16     | 1993-01-01T00:00:00.000Z | 1992-01-01T00:00:00.000Z
+12     | 1992-01-01T00:00:00.000Z | 1991-01-01T00:00:00.000Z
+24     | 1991-01-01T00:00:00.000Z | 1990-01-01T00:00:00.000Z
+26     | 1990-01-01T00:00:00.000Z | 1989-01-01T00:00:00.000Z
```
We don't do double casting, if an explicit casting is already coded in the query, implicit casting won't be added. A copy of the existing `date_nanos.csv` plus some invalid `date_nanos` but valid `date` data is added, and a new schema based on `mapping-date_nanos.json`, with some modifications on the data types is also added, so that we can have some `CsvTests` to show how the implicit casting work on invalid `date_nanos` but valid `date` data.

alex-spies replied:
Thank you for adding the new test data + tests! That's nicely covered, now.

fang-xing-esql commented:
Thanks for reviewing and brainstorming @craigtaverner @alex-spies ! Most of the comments have been addressed, here is a summary of the list of things that have been discussed:

1. The scope of this PR is reduced to do implicit casting for union typed fields with `date` and `date_nanos` only. It makes sense to take it out, as there are a lot of questions with `date_nanos` already, `date_nanos` can be a pilot, so that we can collect feedbacks before supporting the other types.
2. We don't do double casting, if there is explicit casting on union typed fields with `date` and `date_nanos` coded in the query, they won't be cast to date_nanos implicitly.
3. The implicit casting for union typed fields with `date` and `date_nanos` is taken out of the existing `ImplicitCasting` rule, a new rule is added after `ResolveUnionTypes`, to do implicit casting for the union typed fields(without explicit casting) that are not resolved by `ResolveUnionTypes`. We are not always able to replace a union typed field with a conversion function in place, sometimes an `eval` child plan needs to be added for the conversion functions, we need to traverse the plan tree, we can do it inside the `ResolveUnionTypes` rule, if it is preferred, however it will add some complexity to it, that's the main reason that I keep it as a separate rule in `ImplicitCastingForUnionTypedFields`.
4. Enrich doesn't support date_nanos as a match field(checked in [resolveEnrich](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java#L1201)), it is out of the scope of this PR.
5. Lookup join [does not work](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/AbstractLookupService.java#L198) with `date_nanos` join keys yet, it is out of the scope of this PR, join is also a binary plan, in the future if we want to support it, we need some special handlings for binary plans when traversing and transforming the plan.

There are some comments that haven't been addressed yet:
1. Add a warning message when we do implicit casting to `date_nanos`, I think this is a good idea, I haven't added it as I want to clarify with the other folks whether casting to `date_nanos` is the final decision, as `date_nanos` have a smaller range of valid values compared to `date`. I'll reach out to relevant folks to collect feedbacks on which data type to cast to.
2. Performance. Adding implicit casting may affect performance as it may prevent predicates to be pushed down into Lucent. Although today these queries that reference union typed fields fail at Verifier, and have no chance to run, allowing them a run is a step forward, if we can improve their performance by pushing down more into Lucent, that will be ideal. I'll investigate more about the possibility of pushing down more at data nodes.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1780,4 +1790,177 @@ private static LogicalPlan planWithoutSyntheticAttributes(LogicalPlan plan) {
+                            Eval eval = new Eval(u.source(), u.child(), newAliases.values().stream().toList());
+                            p = u.replaceChild(eval);
+                        }
+                        // TODO Lookup join does not work on date_nanos field today, joining on a date_nanos field does not find a match.
```
I've added a bug for date_nanos in LOOKUP JOIN at https://github.com/elastic/elasticsearch/issues/127249, and mentioned that this partial work on implicit casting needs to be done after (or during) that bugfix.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1780,4 +1790,177 @@ private static LogicalPlan planWithoutSyntheticAttributes(LogicalPlan plan) {
+            // if so add a project with eval, so that a not null value can be returned for a union typed field
+            if (logicalPlan.resolved()) {
+                List<Attribute> output = logicalPlan.output();
+                Map<FieldAttribute, Alias> newAliases = Maps.newHashMapWithExpectedSize(output.size());
```
This code is very similar to code above, but that also took into account previous/existing field mappings. Why is that not needed here?

fang-xing-esql replied:
The goal of `castInvalidMappedFieldInFinalOutput` is to add implicit casting to mixed `date`/`date_nanos` fields, that are not referenced by the query(there was no chance to do implicit casting for them if they are not referenced in any commands), however they are in the returned resultsets. For example, `from index*`. It only checks the top level plan's output, it doesn't traverse the entire tree, so it doesn't have a map to keep track of the casted fields map across all plans/subplans.

craigtaverner reviewed (approved):
This looks much nicer to me! I have a few comments and questions, but otherwise I think it is good.

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1780,4 +1790,177 @@ private static LogicalPlan planWithoutSyntheticAttributes(LogicalPlan plan) {
+            // MultiTypeEsField, and don't be double cast here.
+            Map<FieldAttribute, Alias> invalidMappedFieldCasted = new HashMap<>();
+            LogicalPlan transformedPlan = plan.transformUp(LogicalPlan.class, p -> {
+                if (p instanceof UnaryPlan == false && p instanceof LookupJoin == false) {
```
I'll add a comment here for `LookupJoin`, and remove the `instanceof` check, I was attempting to make `date_nanos` fields work as join keys, however it didn't workout easily for me. Thank you for creating a separate issue for it!

fang-xing-esql committed:
push down binary comparisons on date and date nanos union type fields

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1327,13 +1338,35 @@ private BitSet gatherPreAnalysisMetrics(LogicalPlan plan, BitSet b) {
+                    || p instanceof MvExpand
+                    || p instanceof Eval
+                    || p instanceof LookupJoin
+                    || p instanceof Enrich) {
```
> 
If we do the implicit casting in ResolveUnionTypes, without checking the parent/child plan, we may end up cast the union_typed_field twice, which is unnecessary. We may need to add some complexity in this rule to check parent/child relationship, and reduce the amount the casting.

This is a good thing to be paranoid about, but avoiding multiple casts of the same field attributes is a problem that `ResolveUnionTypes` already solved.

For each target type that `union_typed_field` is cast to in the query, `ResolveUnionTypes` adds 1 field attribute `$$union_typed_field$converted_to$target_type` to the `EsRelation`, and all casts of this field to `target_type` are simply replaced by this new field attribute in all expressions. Then later, once field extractions are getting inserted, only one extraction is inserted at the first point where ``$$union_typed_field$converted_to$target_type` is first referenced in the query plan.

alex-spies replied:
I spent some time trying to solve this problem with a smaller change to `ResolveUnionTypes`, only, and while multiple casts would be easily solved this way, I didn't manage to find a solution for the problem that you stated:

> No.2 is that not all of the places where take an InvalidMappedField, take an expression like a ConversionFunction, an Eval plan needs to be added before the plan that references an InvalidMappedField, for example keep/drop/mv_expand/enrich they take identifier or patterns, but they don't take a functionExpression.

Replacing `datetime_nanos_field` by `$$datetime_nanos_field$converted_to$nanos` everywhere in the plan as part of `ResolveUnionTypes` can't work because changing the name messes up the resolution, e.g. when there's a downstream `EVAL` that shadows `datetime_nanos` field.

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

alex-spies commented:
Discussed offline with @fang-xing-esql and @craigtaverner :+1: 

I think the proposed solution should work, it's just a little complex given that it needs to replicate a) some logic present in InsertFieldExtraction (inserting EVALs for implicitly cast fields just before they're used) and b) for passing on fields that users never explicitly dropped or kept into the final output correctly.

I suggested one more approach which we agreed to try before going forward with the Eval-injection approach. The alternative would turn invalid mapped fields of date/nanos type into multi typed fields per default, before any resolution step (multi typed fields in field attributes trigger automatic conversion on field extraction). I threw together a [draft](https://github.com/elastic/elasticsearch/commit/037e878a74ac4fecbe4680054d475d0246fe8bbc) as a commit on top of this PR, which seems to still run most tests fine (all existing union type tests + most tests from this PR).

This approach still needs to special case for the situation where users specifically don't want to convert into nanos, but dates, i.e. when the query has `datetime_nanos_field::datetime` in some expression and the classical union type mechanism should kick in.

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
modify testSuggestedCast as date_nanos and date are casted to date_nanos implicitly

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting

fang-xing-esql committed:
Merge branch 'main' into date-nanos-implicit-casting
